dss/storage/bundles.py

```diff
@@ -22,44 +22,66 @@
     f"^{BUNDLE_PREFIX}/({UUID_PATTERN})(?:\.(" + VERSION_PATTERN + "))?\." + TOMBSTONE_SUFFIX + "$")
 # matches all bundle objects
 DSS_OBJECT_NAME_REGEX = re.compile(
-    f"^{BUNDLE_PREFIX}/({UUID_PATTERN})(?:\.({VERSION_PATTERN}))?(?:\.{TOMBSTONE_SUFFIX})?$")
+    f"^({BUNDLE_PREFIX}|{FILE_PREFIX})/({UUID_PATTERN})(?:\.({VERSION_PATTERN}))?(\.{TOMBSTONE_SUFFIX})?$")
 
 
-def bundle_key_to_bundle_fqid(bundle_key: str) -> str:
-    match = DSS_OBJECT_NAME_REGEX.match(bundle_key)
-    uuid, version = match.groups() if match else (None, None)
-    if uuid and version:
-        return format_bundle_fqid(uuid, version)
-    else:
-        raise ValueError(f"Object name does not contain a valid bundle identifier: {bundle_key}")
+class ObjectIdentifier(namedtuple('ObjectIdentifier', 'uuid version')):
 
+    @classmethod
+    def from_key(cls, bundle_key: str):
+        match = DSS_OBJECT_NAME_REGEX.match(bundle_key)
+        object_type, uuid, version, tombstone_suffix = match.groups() if match else (None, None, None, None)
+        if object_type == FILE_PREFIX:
+            return FileFQID(uuid=uuid, version=version)
+        elif object_type == BUNDLE_PREFIX:
+            if not tombstone_suffix and uuid and version:
+                return BundleFQID(uuid=uuid, version=version)
+            elif tombstone_suffix and uuid:
+                return TombstoneID(uuid=uuid, version=version)
+            else:
+                raise ValueError(f"Object name does not contain a valid bundle identifier: {bundle_key}")
+        else:
+            raise ValueError(f"Object name does not contain a valid bundle identifier: {bundle_key}")
 
-def bundle_fqid_to_uuid_version(bundle_key: str) -> (str, str):
-    match = DSS_BUNDLE_FQID_REGEX.match(bundle_key)
-    uuid, version = match.groups() if match else (None, None)
-    if uuid and version:
-        return uuid, version
-    else:
-        raise ValueError(f"Object name does not contain a valid bundle identifier: {bundle_key}")
+    def is_fully_qualified(self):
+        return self.uuid is not None and self.version is not None
 
+    def to_key(self):
+        return f"{self.prefix}/{self}"
 
-def format_bundle_fqid(uuid: str, version: str) -> str:
-    if UUID_REGEX.match(uuid) and VERSION_REGEX.match(version):
-        return f"{uuid}.{version}"
-    else:
-        raise ValueError(f"Not a valid version regex-version pair: {uuid}, {version}")
+    @property
+    def prefix(self):
+        return NotImplementedError("'prefix' not implemented!")
```

Comment:
https://docs.python.org/3/library/abc.html#abc.abstractproperty

May be overkill

Reply:
Says `@abstractproperty` is deprecated, but I'll add `@abstractmethod`

Reply:
You'd also have to add `metaclass='abc.AbstractBaseClass'` to the `class` line. I wonder if that works with namedtuples.